counsel.el

```diff
@@ -139,6 +139,37 @@ To skip the `executable-find' check, start the string with a space."
      (replace-regexp-in-string          ; Insert dir before any trailing colon.
       "\\(?:: ?\\)?\\'" dir (ivy-state-prompt ivy-last) t t))))
 
+(defun counsel--flatten (tree)
+  "Return a \"flattened\" copy of TREE.
+In other words, return a list of the non-nil terminal nodes, or
+leaves, of the tree of cons cells rooted at TREE.  Leaves in the
+returned list are in the same order as in TREE.
+\(flatten-tree \\='(1 (2 . 3) nil (4 5 (6)) 7))
+=> (1 2 3 4 5 6 7)
+
+This is copied of emacs-27 source code."
+  (let (elems)
+    (while (consp tree)
+      (let ((elem (pop tree)))
+        (while (consp elem)
+          (push (cdr elem) tree)
+          (setq elem (car elem)))
+        (if elem (push elem elems))))
+    (if tree (push tree elems))
+    (nreverse elems)))
+
+(defun counsel--format (formatter &rest args)
+  "Like `format' but FORMATTER can be a list.
+When FORMATTER is a list, only `%s' is replaced with ARGS.
+
+Return a list or string depends on input."
+  (cond
+   ((listp formatter)
+    (counsel--flatten (mapcar
```

Comment:
Why is `counsel--flatten` necessary here?

```el
(counsel--format '("foo" "%s") 'bar) ; => ("foo" bar)
```

The same result can be achieved as follows:

```el
(defun counsel--format (formatter &rest args)
  "Like `format' but FORMATTER can be a list.
When FORMATTER is a list, only `%s' is replaced with ARGS.
Return a list or string depending on input."
  (if (listp formatter)
      (mapcar (lambda (it) (if (equal it "%s") (pop args) it))
              formatter)
    (apply #'format formatter args)))
```

In which case `counsel--flatten` can be deleted.
```

Reply:
Oh, are you using `counsel--flatten` for the case where `(pop args)` returns nil? If so, `counsel--flatten` is overkill; just use `(delq nil ...)` or `cl-mapcan` instead:

```el

(defun counsel--format (formatter &rest args)
  "Like `format' but FORMATTER can be a list.
When FORMATTER is a list, only `%s' is replaced with ARGS.
Return a list or string depending on input."
  (if (listp formatter)
      (cl-mapcan (lambda (fmt)
                   (let ((arg (if (equal fmt "%s") (pop args) fmt)))
                     (and arg (list arg))))
                 formatter)
    (apply #'format formatter args)))
```

Reply:
The flatten is need so I can do something like:
``` elisp
(counsel--format '("foo" "%s") '(bar baz "%s" bae)) ; => ("foo" bar baz "%s" bae)
```
Basically, passing a list to `%s`

Reply:
You still don't need `counsel--flatten` for that. You can use `cl-mapcan`:

```el
(defun counsel--format (formatter &rest args)
  "Like `format' but FORMATTER can be a list.
When FORMATTER is a list, only `%s' is replaced with ARGS.
Return a list or string depending on input."
  (if (listp formatter)
      (cl-mapcan (lambda (fmt)
                   (let ((arg (if (equal fmt "%s") (pop args) fmt)))
                     (cond ((consp arg) (copy-sequence arg))
                           (arg (list arg)))))
                 formatter)
    (apply #'format formatter args)))
```

or `cl-reduce`:

```el
(defun counsel--format (formatter &rest args)
  "Like `format' but FORMATTER can be a list.
When FORMATTER is a list, only `%s' is replaced with ARGS.
Return a list or string depending on input."
  (if (listp formatter)
      (cl-reduce (lambda (acc fmt)
                   (let ((arg (if (equal fmt "%s") (pop args) fmt)))
                     (append acc (if (listp arg) arg (list arg)))))
                 formatter :initial-value ())
    (apply #'format formatter args)))
```

Reply:
That's only solve the problem if it's a flat list passed to `%s`, in case of nested list passed to `%s`, the easiest way is to use flatten.

Reply:
Sure, but can you give a concrete example of where nested lists are necessary in this context? Otherwise it's just overengineering IMO.

Reply:
For example, in the string version, I can concat multiple strings into one and passed to `%s` of `format`. 
However, in the list version, the equivalent to those multiple strings can be mixes between strings or lists. We can concat those elements by either `append`, `nconc` or simply `list`. All of them resulted in an half list with the last element is a `cons` or a nested list.

Reply:
But does that ever give you a nesting of more than 1 level? If so, can't it be avoided by writing the logic in a simpler way? I'm not opposed to maintaining a compatibility shim for a Emacs 27 function, but usually it's indicative that there's a simpler way.

Reply:
I get your points. I'm agree that this can be an overengineering attempt, but from maintenance POV, I think just using a shim to Emacs 27 function is probably the best. 
We:
- Don't have to maintain it much, since it's source code is copied from a maintained source.
- Encourage user to switch to higher Emacs version, which is always a good thing.
- Don't need to care much about the corner cases, since it's a provided function.

Well, mostly I don't want to care about the corner cases that we have to maintain ourselves, which can be quite hard to grab for new contributor/maintainer.

Reply:
Okay, thanks for those points, consider me convinced.